In [173]:
import pandas as pd
import datetime as dt
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn import model_selection
from sklearn.model_selection import train_test_split
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
fie = pd.read_csv("../Family Income and Expenditure.csv")

## 1. รายชื่อกลุ่ม 

- นายกลวัชร เสถียรเสรีสุข  6210450016
- นายเกียรติกุล ทองอุดม    6210450032
- นานชาญวิทย์ สายสิญจน์  6210450521
- น.ส.ณัฎฐกานต์ อรสุนทร   6210451144
- น.ส.ธันวพร ธรรมเจริญพงศ์ 6210451225

## 2. รายละเอียดข้อมูล และการทำความสะอาดข้อมูล ##

**คำอธิบายข้อมูล**
- เป็นการเก็บข้อมูลที่ได้จากการสำรวจรายได้และรายจ่ายของแต่ละครอบครัว ในฟิลิปปินส์ มีประมาณ 40,000 กว่าครัวเรือน โดยเก็บข้อมูลทุกๆ 3 ปี

**แหล่งที่มาของข้อมูล**
- https://www.kaggle.com/grosvenpaul/family-income-and-expenditure

**เก็บรวบรวมโดย**
- Francis Paul Flores

**รายละเอียดคอลัมน์(สนใจแค่คอลัมน์ที่เกี่ยวกับรายรับรายจ่าย และข้อมูลของครอบครัว)**
1. Total Household Income - รายได้ทั้งหมดของแต่ละครัวเรือน type(int)
2. Region - ภูมิภาคของแต่ละครัวเรือน type(object)
3. Total Food Expenditure - รายจ่ายทั้งหมดที่ใช้กับอาหาร type(int)
4. Main Source of Income - แหล่งรายได้หลัก type(object)
5. Agricultural Household indicator - ตัวบ่งชี้ครัวเรือนเกษตร type(int)
6. Bread and Cereals Expenditure - รายจ่ายที่ใช้กับขนมปังและธัญพืช type(int)
7. Total Rice Expenditure - รายจ่ายทั้งหมดที่ใช้กับข้าวtype(int)
8. Meat Expenditure - รายจ่ายทั้งหมดที่ใช้กับเนื้อ type(int)
9. Total Fish and marine products Expenditure - รายจ่ายทั้งหมดที่ใช้กับปลาและภลิตภัณฑ์ที่ทำจากทะเล type(int)
10. Fruit Expenditure - รายจ่ายที่ใช้กับผลไม้ type(int)
11. Vegetables Expenditure - รายจ่ายที่ใช้กับผัก type(int)
12. Restaurant and hotels Expenditure - รายจ่ายที่ใช้กับร้านอาหารและโรงแรม type(int)
13. Alcoholic Beverages Expenditure - รายจ่ายที่ใช้กับเครื่องดื่มแอลกอฮอล์ type(int)
14. Tobacco Expenditure - ค่าใช้จ่ายที่ใช้กับยาสูบ type(int)
15. Clothing, Footwear and Other Wear Expenditure - รายจ่ายที่ใช้กับเสื้อผ้า รองเท้า และเครื่องแต่งกายอื่นๆ type(int)
16. Housing and water Expenditure - ค่าใช้จ่ายที่เกี่ยวกับค่าน้ำและค่าไฟ type(int)
17. Imputed House Rental Value - ค่าเช่าบ้าน type(int)
18. Medical Care Expenditure - รายจ่ายที่ใช้กับค่ารักษาพยาบาล type(int)
19. Transportation Expenditure - รายจ่ายที่ใช้กับการเดินทาง type(int)
20. Communication Expenditure - ค่าใช้จ่ายในกาสื่อสาร type(int)
21. Education Expenditure - รายจ่ายที่ใช้กับการษึกษาเล่าเรียน type(int)
22. Miscellaneous Goods and Services Expenditure - รายจ่ายที่ใช้กับสินค้าและบริการเบ็ดเตล็ด type(int)
23. Special Occasions Expenditure - รายจ่ายที่ใช้ในโอกาสพิเศษ type(int)
24. Crop Farming and Gardening expenses - รายจ่ายที่ใช้กับการปลูกพืชและการทำสวน type(int)
25. Total Income from Entrepreneurial Acitivites - รายได้รวมจากกิจกรรมผู้ประกอบการ type(int)
26. Household Head Sex - เพศของหัวหน้าครั้วเรือน type(object)
27. Household Head Age - อายุของหัวหน้าครัวเรือน type(int)
28. Household Head Marital Status - สถานภาพการสมรสของหัวหน้าครัวเรือน type(object)
29. Household Head Highest Grade Completed - การจบการศึกษาของหัวหน้าครัวเรือน type(object)
30. Household Head Job or Business Indicator - งานหรือธุรกิจของหัวหน้าครัวเรือน type(object)
31. Household Head Occupation - อาชัพของหัวหน้าครอบครัว type(object)
32. Household Head Class of Worker	type(object)
33. Type of Household - ประเภทของครัวเรือน type(object)
34. Total Number of Family members -  จำนวนสมาชิกของแต่ละครัวเรือน type(int)

In [2]:
fie

Total Household Income              Region  Total Food Expenditure  \
0                      480332                 CAR                  117848   
1                      198235                 CAR                   67766   
2                       82785                 CAR                   61609   
3                      107589                 CAR                   78189   
4                      189322                 CAR                   94625   
...                       ...                 ...                     ...   
41539                  119773  XII - SOCCSKSARGEN                   44875   
41540                  137320  XII - SOCCSKSARGEN                   31157   
41541                  133171  XII - SOCCSKSARGEN                   45882   
41542                  129500  XII - SOCCSKSARGEN                   81416   
41543                  128598  XII - SOCCSKSARGEN                   78195   

            Main Source of Income  Agricultural Household indicator  \
0                   Wage/Salaries                                 0   
1                   Wage/Salaries                                 0   
2                   Wage/Salaries                                 1   
3                   Wage/Salaries                                 0   
4                   Wage/Salaries                                 0   
...                           ...                               ...   
41539  Enterpreneurial Activities                                 1   
41540  Enterpreneurial Activities                                 1   
41541  Enterpreneurial Activities                                 2   
41542  Enterpreneurial Activities                                 1   
41543  Enterpreneurial Activities                                 1   

       Bread and Cereals Expenditure  Total Rice Expenditure  \
0                              42140                   38300   
1                              17329                   13008   
2                              34182                   32001   
3                              34030                   28659   
4                              34820                   30167   
...                              ...                     ...   
41539                          23675                   21542   
41540                           2691                    1273   
41541                          28646                   27339   
41542                          29996                   26655   
41543                          43485                   41205   

       Meat Expenditure  Total Fish and  marine products Expenditure  \
0                 24676                                        16806   
1                 17434                                        11073   
2                  7783                                         2590   
3                 10914                                        10812   
4                 18391                                        11309   
...                 ...                                          ...   
41539              1476                                         6120   
41540              1886                                         4386   
41541               480                                         4796   
41542              2359                                        17730   
41543              1985                                         7735   

       Fruit Expenditure  ...  Number of Refrigerator/Freezer  \
0                   3325  ...                               1   
1                   2035  ...                               0   
2                   1730  ...                               0   
3                    690  ...                               0   
4                   1395  ...                               1   
...                  ...  ...                             ...   
41539               1632  ...                               0   
41540               1840  ...                               0   
41541     

In [3]:
fie.columns

Index(['Total Household Income', 'Region', 'Total Food Expenditure',
       'Main Source of Income', 'Agricultural Household indicator',
       'Bread and Cereals Expenditure', 'Total Rice Expenditure',
       'Meat Expenditure', 'Total Fish and  marine products Expenditure',
       'Fruit Expenditure', 'Vegetables Expenditure',
       'Restaurant and hotels Expenditure', 'Alcoholic Beverages Expenditure',
       'Tobacco Expenditure', 'Clothing, Footwear and Other Wear Expenditure',
       'Housing and water Expenditure', 'Imputed House Rental Value',
       'Medical Care Expenditure', 'Transportation Expenditure',
       'Communication Expenditure', 'Education Expenditure',
       'Miscellaneous Goods and Services Expenditure',
       'Special Occasions Expenditure', 'Crop Farming and Gardening expenses',
       'Total Income from Entrepreneurial Acitivites', 'Household Head Sex',
       'Household Head Age', 'Household Head Marital Status',
       'Household Head Highest Grade Compl

### ปรับเปลี่ยนค่า null 

In [4]:
fie.loc[fie["Household Head Occupation"].isnull(),["Household Head Occupation"]] = 'Other'
fie.loc[fie["Household Head Class of Worker"].isnull(),["Household Head Class of Worker"]] = 'Other'
fie

Total Household Income              Region  Total Food Expenditure  \
0                      480332                 CAR                  117848   
1                      198235                 CAR                   67766   
2                       82785                 CAR                   61609   
3                      107589                 CAR                   78189   
4                      189322                 CAR                   94625   
...                       ...                 ...                     ...   
41539                  119773  XII - SOCCSKSARGEN                   44875   
41540                  137320  XII - SOCCSKSARGEN                   31157   
41541                  133171  XII - SOCCSKSARGEN                   45882   
41542                  129500  XII - SOCCSKSARGEN                   81416   
41543                  128598  XII - SOCCSKSARGEN                   78195   

            Main Source of Income  Agricultural Household indicator  \
0                   Wage/Salaries                                 0   
1                   Wage/Salaries                                 0   
2                   Wage/Salaries                                 1   
3                   Wage/Salaries                                 0   
4                   Wage/Salaries                                 0   
...                           ...                               ...   
41539  Enterpreneurial Activities                                 1   
41540  Enterpreneurial Activities                                 1   
41541  Enterpreneurial Activities                                 2   
41542  Enterpreneurial Activities                                 1   
41543  Enterpreneurial Activities                                 1   

       Bread and Cereals Expenditure  Total Rice Expenditure  \
0                              42140                   38300   
1                              17329                   13008   
2                              34182                   32001   
3                              34030                   28659   
4                              34820                   30167   
...                              ...                     ...   
41539                          23675                   21542   
41540                           2691                    1273   
41541                          28646                   27339   
41542                          29996                   26655   
41543                          43485                   41205   

       Meat Expenditure  Total Fish and  marine products Expenditure  \
0                 24676                                        16806   
1                 17434                                        11073   
2                  7783                                         2590   
3                 10914                                        10812   
4                 18391                                        11309   
...                 ...                                          ...   
41539              1476                                         6120   
41540              1886                                         4386   
41541               480                                         4796   
41542              2359                                        17730   
41543              1985                                         7735   

       Fruit Expenditure  ...  Number of Refrigerator/Freezer  \
0                   3325  ...                               1   
1                   2035  ...                               0   
2                   1730  ...                               0   
3                    690  ...                               0   
4                   1395  ...                               1   
...                  ...  ...                             ...   
41539               1632  ...                               0   
41540               1840  ...                               0   
41541     

In [5]:
expenditure_col= []
for i in fie.columns:
    if ("Expenditure" in i) :
        expenditure_col.append(i)
expenditure_col

['Total Food Expenditure',
 'Bread and Cereals Expenditure',
 'Total Rice Expenditure',
 'Meat Expenditure',
 'Total Fish and  marine products Expenditure',
 'Fruit Expenditure',
 'Vegetables Expenditure',
 'Restaurant and hotels Expenditure',
 'Alcoholic Beverages Expenditure',
 'Tobacco Expenditure',
 'Clothing, Footwear and Other Wear Expenditure',
 'Housing and water Expenditure',
 'Medical Care Expenditure',
 'Transportation Expenditure',
 'Communication Expenditure',
 'Education Expenditure',
 'Miscellaneous Goods and Services Expenditure',
 'Special Occasions Expenditure']

## 3. โจทย์การวิเคราะห์เชิงลึก - ในแต่ละครอบครัวมีการใช้จ่ายค่าอาหารและค่าวัตถุดิบอย่างไร

### จุดมุ่งหมายของโมเดลต่างๆ
- เพื่อทำนายค่าใช้จ่ายในด้านของอาหารและวัตถุดิบในการทำอาหาร
- นำมาประกอบการตัดสินใจหรือวางแผนในการใช้เงิน
- เพื่อเป็นแนวทางแต่ละครัวเรือนมีความจำเป็นในวัตถุดิบประเภทไหนมาก อาจนำมาใช้ในการลงทุนได้
- 

#### เลือกตัวแปรต้นที่อาจส่งผลต่อค่าอาหารและวัตถุดิบทั้งหมด
- Total Household Income
- Total Number of Family members
- Members with age less than 5 year old
- Members with age 5 - 17 years old
- Total number of family members employed

### 3.1 สร้างโมเดลเพื่อทำนายค่าอาหาร(Total Food Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [366]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Total Food Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [367]:
tf = LinearRegression()
tf.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [368]:
tf.score(X_test, y_test)

0.5976429854887741

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าอาหาร เท่ากับ 59.76% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 59.76%

### 3.2 สร้างโมเดลเพื่อทำนายค่าวัตถุดิบจำพวกขนมปัง(Bread and Cereals Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [375]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Bread and Cereals Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [376]:
bce = LinearRegression()
bce.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [377]:
bce.score(X_test, y_test)

0.4724754055446291

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

### 3.3 สร้างโมเดลเพื่อทำนายค่าข้าว(Total Rice Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [387]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Total Rice Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [388]:
tre = LinearRegression()
tre.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [389]:
tre.score(X_test, y_test)

0.36266083019266704

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

### 3.4 สร้างโมเดลเพื่อทำนายค่าวัตถุดิบจำพวกเนื้อสัตว์(Meat Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [390]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Meat Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [391]:
me = LinearRegression()
me.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [392]:
me.score(X_test, y_test)

0.3740911326086904

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

### 3.5 สร้างโมเดลเพื่อทำนายค่าวัตถุดิบทางทะเล(Total Fish and  marine products Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [399]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Total Fish and  marine products Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [400]:
tfmpe = LinearRegression()
tfmpe.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [401]:
tfmpe.score(X_test, y_test)

0.2369889448428153

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

### 3.6 สร้างโมเดลเพื่อทำนายค่าวัตถุดิบจำพวกผลไม้(Fruit Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [402]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Fruit Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [403]:
fe = LinearRegression()
fe.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [404]:
fe.score(X_test, y_test)

0.26059845733804976

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

### 3.7 สร้างโมเดลเพื่อทำนายค่าวัตถุดิบจำพวกผัก(Vegetables Expenditure) โดยใช้ Regression

#### Splitting data into training and testing sets - สร้างชุดทดลองและชุดทดสอบ

In [405]:
X = fie.loc[:,['Total Household Income','Total Number of Family members','Members with age less than 5 year old',
               'Members with age 5 - 17 years old','Total number of family members employed']]
y = fie["Vegetables Expenditure"]
X_train, X_test, y_train, y_test = train_test_split(X, y)

#### เรียกใช้ Linear Regression แล้วนำข้อมูลที่ split ไว้มา fit

In [406]:
ve = LinearRegression()
ve.fit(X_train, y_train)

LinearRegression()

#### วัดประสิทธิภาพของโมเดลการทำนายโดยคำนวณค่า $R^2$ จากฟังก์่น score()

In [407]:
ve.score(X_test, y_test)

0.2108493281241598

จะได้ว่าความแปรปรวนที่ลดลงไปเมื่อใช้โมเดลนี้ทำนายค่าวัตถุดิบจำพวกขนมปัง เท่ากับ 47.24% กล่าวได้ว่าโมเดลการทำนายมีประสิทธิภาพ 47.24%

## 4. อธิบายสรุปผล

จะได้ว่าโมเดลต่างๆที่ได้ออกมานั้นบางโมเดลได้ score ค่อนข้างดี แต่ก็มีบางโมเดลที่ score ยังไม่น่าพอใจ เพราะไม่มีตัวแปรต้นที่ส่งผลหรือมีอิทธิพลมากจริงๆกับตัวแปรตาม